alf/environments/metadrive/map_perception.py

```diff
@@ -0,0 +1,162 @@
+        """Initialize by creating the polylines for all the map objects and navigation.
+
+        """
+        polylines = []
+
+        # 1. Collect all the lane boundaries. Each of the lane boundaries will
+        # be divided into a set of polylines. Note that ``Polyline.from_lane()``
+        # returns a Polyline instance that represents a batch of polylines.
+        for _from in road_network.graph.keys():
+            for _to in road_network.graph[_from].keys():
+                # Between two road network nodes _from and _to, there can be
+                # several parallel lanes.
+                lanes = road_network.graph[_from][_to]
+
+                # Now extracting polylines from the lane boundaries.
+                for lane in lanes:
+                    # Since adjacent lanes share boundaries, we are extracting 1
+                    # lane boundaries from each lane, unless it is the last
+                    # lane, where we extract its boundaries of both sides.
+                    num_sides = 2 if lane is lanes[-1] else 1
+                    for side in range(num_sides):
+                        # Ignore lane boundaries with a type of NONE, because
+                        # that means in the real world there will be no line
+                        # drawn on the ground for them, even though logically
+                        # they exist.
+                        if lane.line_types[side] == LineType.NONE:
+                            continue
+                        category = self._category_encoder.encode_line_type(
+                            lane.line_types[side])
+                        offset = -0.5 if side == 0 else 0.5
+                        polylines.append(
+                            Polyline.from_lane(lane, offset, category,
+                                               self._segment_resolution,
+                                               self._polyline_size))
+
+        # 2. Collect all the navigation lines. They are actually center lines of
+        # selected lanes.
+        for i in range(len(navigation.checkpoints) - 1):
+            _from = navigation.checkpoints[i]
+            _to = navigation.checkpoints[i + 1]
+            lanes = road_network.graph[_from][_to]
+            for lane in lanes:
+                category = self._category_encoder.encode_navigation()
+                polylines.append(
+                    Polyline.from_lane(lane, 0.0, category,
+                                       self._segment_resolution,
+                                       self._polyline_size))
+
+        # 3. Now polylines is a list of Polyline instances, where each Polyline
+        # instance corresponds to a batch of polylines. Here we flatten it so
+        # that it becomes a big batch of polylines within one single Polyline
+        # instance, which gets stored in self._polylines.
+        num_polylines = functools.reduce(lambda s, pl: s + pl.point.shape[0],
```

Comment:
Perhaps the following is more readable:
```python
    num_polylines = sum(pl.point.shape[0] for pl in polylines)
```

Reply:
Alas, I did not realize python has builtin `sum`. Thanks!